README.md
# Fake Apache/Nginx Log Generator

This script generates a boatload of fake apache logs very quickly. Its useful for generating fake workloads for [data ingest](http://github.com/streamsets/datacollector) and/or analytics applications.

It can write log lines to console, to log files or directly to gzip files.

It utilizes the excellent [Faker](https://github.com/joke2k/faker/) library to generate realistic ip's, URI's etc.

***

## Basic Usage

Generate a single log line to STDOUT

```bash
$ python apache-fake-log-gen.py  
```

Generate 100 log lines into a .log file

```bash
$ python apache-fake-log-gen.py -n 100 -o LOG 
```

Generate 100 log lines into a .gz file

```bash
$ python apache-fake-log-gen.py -n 100 -o GZ 
```

Infinite log file generation (useful for testing File Tail Readers)

```bash
$ python apache-fake-log-gen.py -n 0 -o LOG 
```

Prefix the output filename

```bash
$ python apache-fake-log-gen.py -n 100 -o LOG -p WEB1
```

Outfile name and directory set and maximum delay between log line writes

```bash
$ python apache-fake-log-gen.py -n 10 -o LOG -o logs --filename apache-access --max-dealy 1000 
```

Detailed help
```
$ python apache-fake-log-gen.py -h
usage: apache-fake-log-gen.py [-h] [--output {LOG,GZ,CONSOLE}]
                              [--num NUM_LINES] [--prefix FILE_PREFIX]

Fake Apache Log Generator

optional arguments:
  -h, --help            show this help message and exit
  --output {LOG,GZ,CONSOLE}, -o {LOG,GZ,CONSOLE}
                        Write to a Log file, a gzip file or to STDOUT
  --num NUM_LINES, -n NUM_LINES
                        Number of lines to generate (0 for infinite)
  --prefix FILE_PREFIX, -p FILE_PREFIX
                        Prefix the output file name
```


## Requirements
* Python 2.7
* ```pip install -r requirements.txt```

## License
This script is released under the [Apache version 2](LICENSE) license.


history.txt
    1  netstat -lt
    2  apt-get install net-tools
    3  apt-get update
    4  sudo apt-get update
    5  sudo apt-get upgrade
    6  docer version
    7  docker version
    8  clear
    9  l
   10  netstat -lt
   11  sudo apt-get install net-tools
   12  netstat -lt
   13  sudo netstat -lt
   14  exit
   15  sudo netstat -lt
   16  l
   17  ls
   18  cd ~
   19  ls -al
   20  cd /
   21  cd vagrant/
   22  ls -al
   23  sudo apt-ge install docker-compose
   24  sudo apt-get install docker-compose
   25  docker-compose up

sudo apt-get install apt-transport-https ca-certificates curl software-properties-common -y
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
sudo apt-key fingerprint 0EBFCD88
sudo add-apt-repository    "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
    $(lsb_release -cs) \
    edge"
sudo apt-get remove docker docker-engine -y
sudo apt-get update
sudo apt-get install docker-ce -y
sudo apt-get install python-pip -y
sudo pip install --upgrade docker-compose

sudo groupadd docker
sudo gpasswd -a ${USER} docker
sudo service docker restart

   35  sudo apt-get install docker
   36  sudo apt-get remove docker docker-engine
   37  sudo apt-get install docker-ce
   38  sudo apt-get install docker.io
   39  sudo apt-get install docker-compose
   40  sudo apt-get upgrade
   41  sudo apt-get install docker-ce
   42  sudo apt-get search docker
   43  apt-cache search docker
   44  sudo apt-get update
   45  exit
   46  sudo apt-get install docker-ce
   47  sudo add-apt-repository    "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   48     $(lsb_release -cs) \
   49     edge"
   50  sudo apt-get update
   51  sudo apt-get install docker-ce
   52  docker install docker-compose
   53  sudo apt-get install docker-compose
   54  sudo apt-get upgrade 
   55  sudo apt-get autoremvoe
   56  sudo apt-get autoremove
   57  docker-compose version
   58  cd /vagrant/
   59  docker-compose up
   60  sudo pip install docker-compose
   62  sudo pip install docker-compose
   63  sudo pip upgrade docker-compose
   64  sudo pip install --upgrade docker-compose
   65  l
   66  docker-compose up
   67  docker-compose version
   68  apt-get uninstall docker-compos
   69  sudo apt-get remove docker-compose
   70  docker-compose 
   72  pyton
   73  python
   74  vim ~/.bashrc 
   75  sudo apt-get install vim
   76  vim ~/.bashrc 
   77  source ~/.bashrc 
   78  docker-compose 
   79  docker-compose up
   80  pip uninstall docker-py; pip uninstall docker; pip install docker
   81  docker-compose up
   82  sudo service start docker
   83  sudo service docker start
   84  docker-compose up
   85  docker
   86  docker ps
   87  sudo docker-compose up
   88  l
   89  ls -al
   90  sudo docker-compose up

   94  docker-compose up
   95  docker ps
   96  docker
   97  docker ps
   98  sudo service docker restart
   99  exit
  100  docker ps
  101  cd /vagrant/
  102  docker-compose up
  103  chown 0 filebeat.yml 
  104  docker-compose up
  105  exit
  106  history 
  107  history : /vagrant/history.txt
  108  history > /vagrant/history.txt
  109  whoami
  110  sudo history > /vagrant/history.txt
  111  cd /vagrant/
  112  ls
  113  ls -al
  114  history > history.txt
  115  exit
  116  cd /vagrant/
  117  ls
  118  history > history.txt
  119  exit
  120  cd /vagrant/
  121  touch hello.txt
  122  ls -al
  123  man mount.prl_fs
  124  man
  125  cd /vagrant/
  126  ls -al
  127  touch hello.txt
  128  sudo touch hello.txt
  129  whoami
  130  sudo su root
  131  exit
  132  cd /vagrant/
  133  touch hello.txt
  134  l
  135  history > history.txt 


log_write_sleep.py
from random import randint
import os


def write_log_directory(output_dir = ''):
    if output_dir == '':
        return output_dir

    if not output_dir.endswith(os.sep):
        output_dir += os.sep

    if os.path.isabs(output_dir) and not os.path.exists(output_dir):
        print('No such directory: %s. Please create it will fallback to current directory', output_dir)
        return ''

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    return output_dir


def write_sleep(min_delay = 0, max_dealy = 0):
    if min_delay == 0 and max_dealy == 0:
        return 0

    max_dealy = valid_delay(max_dealy)
    min_delay = valid_delay(min_delay)
    if max_dealy == 0 and min_delay != 0:
        max_dealy = + 10

    return randint(min_delay, max_dealy)


def valid_delay(num):
    return num if num > 0 else 0


apache-fake-log-gen.py
#!/usr/bin/python
import time
import datetime
import pytz
import numpy
import random
import gzip
import zipfile
import sys
import argparse
from faker import Faker
from random import randrange
from tzlocal import get_localzone
import log_write_sleep

local = get_localzone()


# todo:
# allow writing different patterns (Common Log, Apache Error log etc)
# log rotation


class switch(object):
    def __init__(self, value):
        self.value = value
        self.fall = False

    def __iter__(self):
        """Return the match method once, then stop"""
        yield self.match
        raise StopIteration

    def match(self, *args):
        """Indicate whether or not to enter a case suite"""
        if self.fall or not args:
            return True
        elif self.value in args:  # changed for v1.5, see below
            self.fall = True
            return True
        else:
            return False


parser = argparse.ArgumentParser(__file__, description="Fake Apache Log Generator")
parser.add_argument("--output", "-o", dest='output_type', help="Write to a Log file, a gzip file or to STDOUT",
                    choices=['LOG', 'GZ', 'CONSOLE'])
parser.add_argument("--num", "-n", dest='num_lines', help="Number of lines to generate (0 for infinite)", type=int,
                    default=1)
parser.add_argument("--prefix", "-p", dest='file_prefix', help="Prefix the output file name", type=str)
parser.add_argument("--sleep", "-s", help="Sleep this long between lines (in seconds)", default=0.0, type=float)

parser.add_argument('--output-dir', '-d', help='Output directory for log files', default="", type=str)
parser.add_argument('--filename', '-f', help='Log file name', default="", type=str)
parser.add_argument('--min-delay', help='Minimum delay between writes in milliseconds', default=0, type=int)
parser.add_argument('--max-delay', help='Maximum delay between writes in milliseconds', default=0, type=int)

args = parser.parse_args()

log_lines = args.num_lines
file_prefix = args.file_prefix
output_type = args.output_type
min_write_delay = args.min_delay
max_write_delay = args.max_delay
output_filename = args.filename
output_dir = log_write_sleep.write_log_directory(args.output_dir)

faker = Faker()

timestr = time.strftime("%Y%m%d-%H%M%S")
otime = datetime.datetime.now()

if output_filename == '':
    outFileName = 'access_log_' + timestr + '.log' if not file_prefix else file_prefix + '_access_log_' + timestr + '.log'
else:
    outFileName = output_filename + '.log'

outFileName = output_dir + outFileName
print('Output file: %s', outFileName)

for case in switch(output_type):
    if case('LOG'):
        f = open(outFileName, 'a')
        break
    if case('GZ'):
        f = gzip.open(outFileName + '.gz', 'w')
        break
    if case('CONSOLE'): pass
    if case():
        f = sys.stdout

response = ["200", "404", "500", "301"]

verb = ["GET", "POST", "DELETE", "PUT"]

resources = ["/list", "/wp-content", "/wp-admin", "/explore", "/search/tag/list", "/app/main/posts",
             "/posts/posts/explore", "/apps/cart.jsp?appID="]

ualist = [faker.firefox, faker.chrome, faker.safari, faker.internet_explorer, faker.opera]

flag = True
while (flag):
    write_sleep = log_write_sleep.write_sleep(min_write_delay, max_write_delay)
    if write_sleep != 0:
        time.sleep(write_sleep / 1000)

    if args.sleep:
        increment = datetime.timedelta(seconds=args.sleep)
    else:
        increment = datetime.timedelta(seconds=random.randint(30, 300))
    otime += increment

    ip = faker.ipv4()
    dt = otime.strftime('%d/%b/%Y:%H:%M:%S')
    tz = datetime.datetime.now(local).strftime('%z')
    vrb = numpy.random.choice(verb, p=[0.6, 0.1, 0.1, 0.2])

    uri = random.choice(resources)
    if uri.find("apps") > 0:
        uri += repr(random.randint(1000, 10000))

    resp = numpy.random.choice(response, p=[0.9, 0.04, 0.02, 0.04])
    byt = int(random.gauss(5000, 50))
    referer = faker.uri()
    useragent = numpy.random.choice(ualist, p=[0.5, 0.3, 0.1, 0.05, 0.05])()
    f.write('%s - - [%s %s] "%s %s HTTP/1.0" %s %s "%s" "%s"\n' % (ip, dt, tz, vrb, uri, resp, byt, referer, useragent))

    log_lines = log_lines - 1
    flag = False if log_lines == 0 else True
    if args.sleep:
        time.sleep(args.sleep)
